feat: Support for iOS

markvideon opened the pull request:
Addressing #19 , iOS support is based on gamepads_darwin with minor adjustments - e.g. the platform checks refer to relevant versions of iOS. 

Note that the example project as-is will fail due to the published release of `gamepads` 0.1.1 not depending on `gamepads_ios`. Manually adding `gamepads_ios: ^0.1.1` as a dependency to the example project will allow you to test everything links as expected.

luanpotter commented on packages/gamepads_ios/ios/gamepads_ios.podspec:
```diff
@@ -0,0 +1,23 @@
+#
+Pod::Spec.new do |s|
+  s.name             = 'gamepads_ios'
+  s.version          = '0.0.1'
```
should this be 0.1.1?

markvideon replied:
This is actually 0.0.1 on gamepads_darwin too

luanpotter commented on packages/gamepads_ios/ios/gamepads_ios.podspec:
```diff
@@ -0,0 +1,23 @@
+Pod::Spec.new do |s|
+  s.name             = 'gamepads_ios'
+  s.version          = '0.0.1'
+  s.summary          = 'A new Flutter plugin project.'
+  s.description      = <<-DESC
+A new Flutter plugin project.
+                       DESC
+  s.homepage         = 'http://example.com'
+  s.license          = { :file => '../LICENSE' }
+  s.author           = { 'Your Company' => 'email@example.com' }
```
can we update these fields?

luanpotter reviewed (commented):
Thanks for implementing support for iOS!

I have a couple questions / concerns:

* I am not sure if we should commit all these "build" files on the examples/ios build folder; @spydon are you familiar with the best practices here? maybe we could revisit the gitignore or ignore all platform build folders
* I am a bit confused on why this doesn't seem to leverage the darwin implementation? I was expecting either `gamepads_darwin` to be used directly, just as it is with macos (maybe with some preprocessing instructions on top), or a new gamepads_ios package to depend on darwin and reuse most of the code. the way I am seeing the code was duplicated for ios on gamepads_ios? or am I missing something?

markvideon commented:
> Thanks for implementing support for iOS!
> 
> 
> 
> I have a couple questions / concerns:
> 
> 
> 
> * I am not sure if we should commit all these "build" files on the examples/ios build folder; @spydon are you familiar with the best practices here? maybe we could revisit the gitignore or ignore all platform build folders
> 
> * I am a bit confused on why this doesn't seem to leverage the darwin implementation? I was expecting either `gamepads_darwin` to be used directly, just as it is with macos (maybe with some preprocessing instructions on top), or a new gamepads_ios package to depend on darwin and reuse most of the code. the way I am seeing the code was duplicated for ios on gamepads_ios? or am I missing something?

RE: second point - is it possible to do this at this stage (e.g. via symbolic links and some shared Apple folder)? Agree it would be ideal

There's a proposal doc from Flutter team below for the general problem of sharing code between macOS and iOS plugins 

https://docs.google.com/document/d/1JdgBrMXV0a6fnVoooILNTQB8dNCvLloT4D4TBkECQ6c/edit?resourcekey=0-iPL3CzwOjUquabkMiucaHA

spydon commented:
>     * I am not sure if we should commit all these "build" files on the examples/ios build folder; @spydon are you familiar with the best practices here? maybe we could revisit the gitignore or ignore all platform build folders

I agree that we probably should skip them, but since we currently have the other build directories committed I think we can do this clean-up as a later step.

>     * I am a bit confused on why this doesn't seem to leverage the darwin implementation? I was expecting either `gamepads_darwin` to be used directly, just as it is with macos (maybe with some preprocessing instructions on top), or a new gamepads_ios package to depend on darwin and reuse most of the code. the way I am seeing the code was duplicated for ios on gamepads_ios? or am I missing something?

Yeah, I don't think that is possible currently unfortunately. It's definitely not possible with symlinks.

spydon commented:
One dirty option we could do is have the files in one directory and symlink it and then add a pre-publish hook to melos that properly copies over the files before publishing.

markvideon commented:
Updated various fields on the iOS podspec and also updated the macOS podspec (gamepads_darwin) version number so that both are at 0.1.1. 

Regarding the discussion around re-use strategy, I'm open to all options, including leaving duplicate code in-place until there is a solution from the framework.

spydon commented:
We can handle the de-duplication matter in a separate PR I think, if we decide on something.

markvideon commented:
Did a bit of reading on this and it seems like we might be able to create a Swift package with shared code but it would require consumers of the package to add a Swift package repository that we manage.

spydon commented:
> Did a bit of reading on this and it seems like we might be able to create a Swift package with shared code but it would require consumers of the package to add a Swift package repository that we manage. 

That doesn't seem worth it, a lot less convenient for the users.

markvideon commented:
> > Did a bit of reading on this and it seems like we might be able to create a Swift package with shared code but it would require consumers of the package to add a Swift package repository that we manage. 
> 
> That doesn't seem worth it, a lot less convenient for the users.

I agree.

spydon reviewed (approved):
Lgtm! But maybe you can actually remove the build directory from the example already, the PR will be much cleaner that way, since they are like 80% of the files. 😅

markvideon commented:
> Lgtm! But maybe you can actually remove the build directory from the example already, the PR will be much cleaner that way, since they are like 80% of the files. 😅

Do you mean removing example/ios entirely?  Should I do this for all platforms?

spydon commented:
> > Lgtm! But maybe you can actually remove the build directory from the example already, the PR will be much cleaner that way, since they are like 80% of the files. 😅
> 
> Do you mean removing example/ios entirely?

Exactly

> Should I do this for all platforms?

We can do that in a separate PR

luanpotter commented:
can't we do the same thing audioplayers does?

https://github.com/bluefireteam/audioplayers/tree/main/packages/audioplayers_darwin

there is a single darwin package for both ios and macos
